[core] Refactor the build for AoT compilation

jeeyun opened the pull request:
Signed-off-by: Jeeyun Lim <jeeyun.lim@gmail.com>

Fixes #62, #220. 

Here's a refactoring of our project to support AoT compilation. Some notable changes are:
- Code changes to make our components AoT compatible:
   - make all inputs and outputs public
   - removed usage of our animation functions in components; the code is still there but not being used; for now the animation is hard-coded into the components (the ngc compiler couldn't statically analyze the functions passed into the component decorator; there were some suggestions to export them as named function but I couldn't get it working; let me know if anyone is able to get this working)
   - other minor edits to get ngc compiler to compile
- Moved demo components to under app (we previously had it nested under each component which required a lot of renaming and shifting things around to produce the npm packages)
- Reduced some of the "automagic" in our build flow:
   - renamed `src/clarity` to `src/clarity-angular`
   - no more auto re-exporting in `src/clarity-angular/index.ts` in favor of explicitly adding export statements for new components
- Now in dev mode, we no longer inline the templates (had to re-add module.id in most components to support this). this allows for the typescript compilation to not depend on sass compilation on watch tasks. prod mode will inline the template as before.
- Added build step to compile our components with ngc compiler. Our clarity-angular npm package will change to include es2015 compiled code (for AoT compilation) along with `metadata.json` files, original source code in `src` folder (as before), and `clarity-angular.min.js` in commonjs format (as before).
- No more demo npm package.

I left some of the unused gulp tasks commented out. Once we are sure we won't need them, I can delete those files.

jeeyun commented:
Made some additional changes, which I've squashed and merged into one commit:
-  removed the added module.id from clarity components (I couldn't get the components to be happy with webpack in dev mode. I think this issue is related: https://github.com/angular/angular-cli/issues/3576). Demo app and components will still use module.id. Clarity components will always be inlined, even in dev mode.
- now producing additional `clarity-angular.umd.js` file. I left the `clarity-angular.min.js` for now, so that is included in the npm package as before. apps using systemjs will still need to be updated to include the umd file in the config.

youdz commented on build/tasks/publish.js:
```diff
@@ -106,28 +104,30 @@ gulp.task("npm:ui:readme", function () {
  * The deliverables for clarity-angular are:
- *   - the compiled js files in commonJS format
- *   - the minified JS bundle
+ *   - the compiled js files in es2015 format
```
I'm a bit scared of this. I think the standard is ES5 with ES2015 module format only, not the full features. Otherwise you either force Clarity's user to use yet another transpiler to ES5, or they forget about it and their app will only run on a handful of browsers.

For instance, when I look at the AoT documentation, they do this in their `tsconfig-aot.json`:
```javascript
{
  "compilerOptions": {
    "target": "es5",
    "module": "es2015",
    ...
  }
}
```

I'm guessing there is a good chance that's what you're doing, but the comments and commit message should probably be less ambiguous when you say "es2015 format". I know it's a mouthful, but "es5 with es2015 module format" will avoid confusion.

jeeyun replied:
Absolutely. Yes, we are doing ES5 with ES2015 module format only as you guessed. Will update the comments.

youdz commented on build/tasks/publish.js:
```diff
@@ -106,28 +104,30 @@ gulp.task("npm:ui:readme", function () {
     return es.merge(
         gulp.src([
             "dist/bundles/clarity-angular.min.js",
             "dist/bundles/angular1/clarity-angular1.min.js"
```
I think you can remove that. 😄

jeeyun replied:
👍

youdz commented on build/tasks/typescript.js:
```diff
@@ -27,118 +27,109 @@ var runSequence = require('run-sequence');
-    gulp.watch(claritySources, function () {
-        return runSequence('tslint:watch', 'typescript:clarity');
+    gulp.watch(claritySources.concat(clarityHtmlFiles), function () {
+        return runSequence('tslint:clarity:watch', 'typescript:clarity');
```
We could have a separate watcher on HTML that doesn't re-run tslint, if it becomes too slow at some point.

PS: `"tslint:clarity:watch"` is misleading, it made me pause and go look again because I thought our watcher was starting a new watcher every time. Maybe `"tslint:clarity:no-errors"` or something similar would be clearer.

jeeyun replied:
I did separate out the tslint tasks to only run on the files that it's watching (earlier we had one monolithic tslint that ran for each typescript subgroups that the project was watching) so hopefully it's a bit better. The clarity components one is a special case because we inline the html. Of course this won't have to be the case if we opt to publish our html files (see my earlier comment on this).

I agree with the naming; I'll rename the tslint tasks.

youdz commented on gulpfile_old.js:
```diff
@@ -0,0 +1,88 @@
+/*
```
Don't forget to remove this when you're done. 😄

youdz commented on src/app/alert/alert.demo.module.ts:
```diff
@@ -0,0 +1,69 @@
+/*
```
This comment is _not_ something I want you to change right now, but I'm writing it down before I forget.

I was thinking about it this morning, and I think we should make it clearer what these are. I would write them in a separate `src/sandbox/` folder, that would be used by `src/app`. Of course, we'd remove a lot of the demos to stick to a single fully-featured example most of the time. So we'd do this as part of the demo cleanup rather than this PR.

youdz commented on src/app/app.module.ts:
```diff
@@ -24,7 +24,8 @@ import {ClarityModule} from "../clarity/clarity.module";
     ],
-    bootstrap: [ AppComponent ]
+    bootstrap: [ AppComponent ],
+    schemas: [ CUSTOM_ELEMENTS_SCHEMA ]
```
This is not needed since the module already imports ClarityModule. I'd rather you didn't add the schema, and if you have errors we import ClarityModule from the various modules that need it. That way we can make sure it works without schema for most people. If you need help handling these errors, I'd be happy to.

jeeyun replied:
Not sure why I introduced this in the first place. Reverting.

youdz commented on src/clarity-angular/modal/modal.ts:
```diff
@@ -0,0 +1,117 @@
+        :host { display: inline-block; }
+    `],
+    animations: [
+        trigger("fadeDown", [
```
Regarding this, I'm fine pushing with duplicate hard-coded animations for now, as long as we create an issue to fix it before we forget.

youdz commented on src/clarity-angular/wizard/wizard-page.ts:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    ngOnInit() {
+        let projectedTitleHTML: string = this.titleContainer.nativeElement.innerHTML.trim();
```
How does this work with AoT? We're accessing a property of the native element. I thought that was one of the things we couldn't do.

jeeyun replied:
Hmm I wasn't aware of limitations wrt accessing a native element's property in AoT. That being said, I tested the components in AoT, and they are working as expected. Let me know if you have any specific documentation or issue that you saw this in and I can take a look.

youdz commented on src/clarity-angular/wizard/wizard-step.ts:
```diff
@@ -0,0 +1,56 @@
+
+    ngOnInit(): void {
+        // The nav text will be the title of the main section.
+        this.title = this.elementRef.nativeElement.innerText.trim();
```
Same question, how does AoT accept that?

jeeyun commented on build/karma.conf.js:
```diff
@@ -61,7 +61,7 @@ module.exports = function (config) {
         ],
 
-        exclude: [],
+        exclude: [ 'node_modules/**/*spec.js' ],
```
Since we include other js files from `node_modules` in karma, making it `node_modules/**` overrides those include patterns declared earlier, and the tests won't be able to run. I had to introduce this because now we include `@angular/compiler-cli` in our project for AoT, and one of their spec files, will throw an error if we don't exclude (see: https://github.com/angular/angular/issues/13115). I think this is generic and safe enough. Let me know if you still disagree.

jeeyun commented on build/tasks/aot.js:
```diff
@@ -0,0 +1,67 @@
+    ];
+
+    gulp.src(claritySources)
+        .pipe(inlineNg2Template({
```
If we opt to not inline the templates, we need to include our template files (which we haven't been) in our npm packages. Do we want to do that?

jeeyun commented on build/tasks/clean.js:
```diff
@@ -12,8 +12,9 @@ var del = require("del");
+		"aot-compiled/*",
 		"dist/*",
-		"tmp/"
+		"tmp/*"
```
There isn't. Actually let me clean up the clean task to delete all the generated folders themselves.

jeeyun commented:
@youdz: I think I addressed all of your comments. Let me know if I missed any. I will amend the commit message with the details when it's ready for merge.

Regarding a gulp task to test AoT compiled component in our demo app, I think it can be done but would it be ok if I file a separate issue for it? I don't know if I can fit it in for this week's release.

youdz commented on build/tasks/publish.js:
```diff
@@ -112,16 +112,17 @@ gulp.task("npm:ui", ["npm:ui:bundles", "npm:ui:sources", "npm:ui:package", "npm:
  */
 gulp.task("npm:angular:bundles", function () {
     // We have to use 2 streams here because we want the angular 1 bundle to preserve its path.
     return es.merge(
```
Nitpick: We don't need to merge two streams here anymore. We can just `gulp.src()` all at once.

Don't bother updating the review for this, it doesn't even matter if it stays that way, this is really a negligible nitpick.
